ambry-commons/src/main/java/com.github.ambry.commons/HelixNotifier.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.HelixPropertyListener;
+import org.apache.helix.store.HelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * <p>
+ *    A {@link Notifier} implementation using {@link HelixPropertyStore} as the underlying communication mechanism.
+ * Each topic is created as a {@link ZNRecord} under the {@code topicPath}. A message for that topic will
+ * be written to the simple field of the {@link ZNRecord} in the form of {@code "message": "message content"}.
+ * Each {@link ZNRecord} under the {@code topicPath} will be used for a single topic.
+ * </p>
+ * <p>
+ *   If multiple {@link HelixNotifier} running on different node are configured with the same
+ *   {@link HelixPropertyStoreConfig}, they will essentially be in the same topic domain. I.e., {@link TopicListener}s
+ *   can subscribe through any of these {@link HelixNotifier}s for a topic. A message of that topic sent through
+ *   {@link HelixNotifier}-1 on node-1 can be received by all the {@link TopicListener}s on other nodes.
+ * </p>
+ */
+public class HelixNotifier implements Notifier<String> {
+  private static final String TOPIC_PATH = "/topics";
+  private static final String MESSAGE_KEY = "message";
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final Map<TopicListener, HelixPropertyListener> topicListenerToHelixListenerMap = new ConcurrentHashMap<>();
+
+  /**
+   * Constructor.
+   * @param storeConfig The {@link HelixPropertyStoreConfig} that specifies necessary configs to start a HelixNotifier.
+   * @param storeFactory The factory to start a {@link HelixPropertyStore}.
+   */
+  public HelixNotifier(HelixPropertyStoreConfig storeConfig, HelixPropertyStoreFactory<ZNRecord> storeFactory) {
+    if (storeFactory == null) {
+      throw new IllegalArgumentException("storeFactory cannot be null.");
+    }
+    List<String> subscribedPaths = new ArrayList<>();
+    subscribedPaths.add(storeConfig.rootPath + TOPIC_PATH);
+    helixStore = storeFactory.getHelixPropertyStore(storeConfig, subscribedPaths);
+    logger.info("HelixNotifier started, topicPath={}", storeConfig.rootPath + TOPIC_PATH);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Return {@code true} does not guarantee all the {@link TopicListener} will receive the message. It just indicates
+   * the message has been successfully sent out.
+   */
+  @Override
+  public boolean sendMessage(String topic, String message) {
+    if (topic == null) {
+      throw new IllegalArgumentException("topic cannot be null");
+    }
+    if (message == null) {
+      throw new IllegalArgumentException("message cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    ZNRecord record = new ZNRecord(topicPath);
+    record.setSimpleField(MESSAGE_KEY, message);
+    boolean res = helixStore.set(topicPath, record, AccessOption.PERSISTENT);
```

Comment:
Could you briefly explain how this works internally (could not quickly find it out from Helix documentation): basically, since you are doing a `set()` which presumably sets the value in some znode (and not really an append), does Helix make the notifications from within the `set()` method itself (that is, is it blocking)? If not, could we end up losing messages if `set()` is called quickly in succession?